### ISSUE TITLE
Calling from within Python

### ISSUE THREAD

**Issue by fladd** [raimon49/pip-licenses#81]
The documentation currently does not explain how to call pip-licenses from within Python, only how to call it from a command line.

**Comment by raimon49** [comment]
Yes, pip-licenses is developed as a CLI tool.

Does your opinion mean that you want to import pip-licenses as a module and use some of the features in your Python code?

**Comment by fladd** [comment]
Yes, that would be great. Currently I get around it by calling it with subprocess and then parsing the json output, but this seems a bit unnecessary, given that pip-licenses is Python code itself.

**Comment by raimon49** [comment]
I didn't imagine anyone would want to use the scan results in their Python code.

I would consider APIs and documentation that users can handle with structured objects.

Thank you for your valuable opinion, @fladd .

**Comment by sweigert** [comment]
> Yes, that would be great. Currently I get around it by calling it with subprocess and then parsing the json output, but this seems a bit unnecessary, given that pip-licenses is Python code itself.

This is what I am doing to achieve this:
```python
import piplicenses
parser = piplicenses.create_parser()
args = parser.parse_args()
packages = piplicenses.get_packages(args)
for p in packages:
   print(p)
```

**Comment by diazona** [comment]
This would also be useful to work around the lack of "proper" support for multiply-licensed packages - that is, the fact that multiple licenses are shown in a single string separated by delimiters, rather than as some structured representation (like a JSON list, for the JSON output format). If there were a Python API, perhaps it could expose multiple licenses as a Python list.

I don't know how practical this actually is, but it would make a significant difference to my use case.

**Comment by bartfeenstra** [comment]
I'd like to express my support for this feature request. I'm currently using pip-licenses in a unit test to ensure all dependencies' licenses are compatible. A pip-licenses Python API would allow this code to be rewritten a lot shorter and cleaner.

**Comment by pwuertz** [comment]
A pip-licenses Python API would be a great fit for applications bundled with e.g. PyInstaller.

Using `piplicenses` within a PyInstaller-specfile, you could fully integrate the licence collection process right into your application bundling process.

**Comment by stefan6419846** [comment]
I previously used `packages = list(piplicenses.get_packages(from_source=piplicenses.FromArg.MIXED))` to get the packages from within Python without relying on a subprocess.

Due to additional requirements from my side (matching all license files, only querying metadata without files etc.) and the upstream repository receiving corresponding PRs which would never end up in a release, I decided to create my own fork of this tool which has a strict focus on library usage: https://github.com/stefan6419846/pip-licenses-lib

**Comment by Lewik** [comment]
> > Yes, that would be great. Currently I get around it by calling it with subprocess and then parsing the json output, but this seems a bit unnecessary, given that pip-licenses is Python code itself.
> 
> This is what I am doing to achieve this:
> 
> ```python
> import piplicenses
> parser = piplicenses.create_parser()
> args = parser.parse_args()
> packages = piplicenses.get_packages(args)
> for p in packages:
>    print(p)
> ```
> 
> If you need options, you could provide them in the call to `parse_args`.

`parser.parse_args()` without arguments will `use sys.argv`
pass `[]`

**Comment by reactive-firewall** [comment]
# A long overdue update on GHI-81 (calling `piplicenses` from within python)

> [!NOTE]
> The plan is to start supporting `import piplicenses` and direct calls to `piplicenses.main(*args, **kwargs)` by version 6.2 (hopefully with a preview sooner in the version 6 series TBD)

## Early details for transparency

So thanks to work from GHI #316, there is an internal module `piplicenses.cli.config` that will eventually split off a class called `Configuration` (I say "eventually" because at the moment it is just the v5.5.5 `CustomNamespace` refactored by that work from GHI-316). The `Configuration` is intended to be transformed into a more straight-forward `dataclass`/`TypedDict` (TBD which) where the results of parsing command-line arguments (which is why it is currently in the new internal `.cli.config` module) as a runtime data instance (e.g. the configured context previously only available from `argparse`).

The plan is to start supporting `import piplicenses` and direct calls to `piplicenses.main(*args, **kwargs)` by version 6.2 (hopefully with a preview sooner in the version 6 series TBD)

However, once the `Configuration` API is more mature, I plan on extending _select_ functions to the public API (e.g. what one gets from a top-level `import piplicenses`) for things like targeted auditing of licenses (think `pip-licenses --fail-on ... -p targPkg` stuff from python calls e.g., `piplicenses.audit_pkg_by_name(name='targPkg', config=your_configuration)`)

This will be a bit different from what the built-in `importlib` Python module provides (btw, `pip-licenses` get's it's list from essentially: `importlib.metadata.distributions(path=' '.join(sys.path).split())` (for all the overhead see https://github.com/raimon49/pip-licenses/blob/0a154eb586ae23e2549e4f41de8a82766fea9a52/pip-licenses/core.py#L579 for implementation))

## API scope (it won't be everything)

Regarding internal functions like `get_packages()` being supported is still TBD however. And I mean more than just effort allocation in this case. Quoting @Lewik and @sweigert's conversation:

> > > Yes, that would be great. Currently I get around it by calling it with subprocess and then parsing the json output, but this seems a bit unnecessary, given that pip-licenses is Python code itself.
> > 
> > 
> > This is what I am doing to achieve this:
> > import piplicenses
> > parser = piplicenses.create_parser()
> > args = parser.parse_args()
> > packages = piplicenses.get_packages(args)
> > for p in packages:
> >    print(p)
> > 
> > If you need options, you could provide them in the call to `parse_args`.
> 
> `parser.parse_args()` without arguments will `use sys.argv` pass `[]`

This starts to beg the question of what is/isn't in scope for `pip-licenses`' API. Not to nitpick the example (:speak_no_evil:) but I feel you don't even need `pip-licenses` for just the packages or their metadata; Python standard library already provides all of this. E.g.:

```python
import sys  # for sys.path()

# optional for typing
from email.message import Message

from importlib import metadata
from importlib.metadata import (
    Distribution,
    PackagePath,
)
# or just
import importlib
# then just print list (e.g., by name)
packages = importlib.metadata.distributions(path=' '.join(sys.path).split())
for p in packages:
    print(p.metadata["name"])

```

So I find that use-case to be kinda unnecessary to support (and to possibly violate [AHA](https://web.archive.org/web/20250413022457/https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction)), or even export from the module, as the scope for `pip-licenses` is (and has been) more narrowly focused on licenses (and depends on Python to provide the packages).

E.g., `pip-licenses` is focused on providing a better alternative to just:

```console
$ python
>>> import importlib
>>> packages = importlib.metadata.distributions(path=' '.join(sys.path).split())
>>> for p in packages:
...     print(p.metadata.get_all("License", None))
```

## Next Steps

So in-order to close out this GHI, documentation is still needed:
 * notably the usage and API of the main entry point
   * including basics (e.g. call conventions and implementation notes)
   * and reproducible API usage examples

:bow: thanks for your continued interest in this.

**Comment by reactive-firewall** [comment]
> I'm currently using pip-licenses in a unit test to ensure all dependencies' licenses are compatible.

@bartfeenstra, any chance those unit-tests are/could be public? With your permission I'd like to document the migration path for that use-case, and knowing how you previously checked in a unit-test would be helpful.
